src/Microsoft.AspNetCore.Server.Kestrel/Internal/Http/FrameHeaders.cs

```diff
@@ -211,25 +213,80 @@ IEnumerator IEnumerable.GetEnumerator()
             if (headerCharacters != null)
             {
-                foreach (var ch in headerCharacters)
+                fixed (char* header = headerCharacters)
                 {
-                    if (ch < 0x20 || ch > 0x7E)
+                    // offsets and lengths are handled in byte* sizes
+                    var pHeader = (byte*)header;
+                    var offset = 0;
+                    var length = headerCharacters.Length * 2;
+
+                    if (Vector.IsHardwareAccelerated && Vector<byte>.Count <= length)
+                    {
+                        var vSub = Vector.AsVectorUInt16(new Vector<uint>(0x00200020u));
+                        // 0x7e as highest ascii (don't include DEL)
+                        // 0x20 as lowerest ascii (space)
+                        var vTest = Vector.AsVectorUInt16(new Vector<uint>(0x007e007eu - 0x00200020u));
+
+                        do
+                        {
+                            var stringVector = Unsafe.Read<Vector<ushort>>(pHeader + offset);
+                            offset += Vector<byte>.Count;
+                            if (Vector.GreaterThanAny(stringVector - vSub, vTest))
+                            {
+                                ThrowInvalidHeaderCharacter(pHeader + offset, Vector<byte>.Count);
+                            }
+                        } while (offset + Vector<byte>.Count <= length);
+                    }
+
+                    // Non-vector testing:
+                    // Flag > 0x007f => Use value directly, already flagged
+                    // Flag 0x7f     => Add 0x0001 to each char so DEL (0x7f) will set a high bit
+                    // Flag < 0x20   => Sub 0x0020 from each char so high bit will be set in previous char bit
+                    // Bitwise | or the above three together
+                    // Bitwise & and each char with 0xff80; result should be 0 if all tests pass
+                    if (offset + sizeof(ulong) <= length)
+                    {
+                        do
+                        {
+                            var stringUlong = (ulong*)(pHeader + offset);
+                            offset += sizeof(ulong);
+                            if (((*stringUlong | (*stringUlong + 0x0001000100010001UL) | (*stringUlong - 0x0020002000200020UL)) & 0xff80ff80ff80ff80UL) != 0)
+                            {
+                                ThrowInvalidHeaderCharacter(pHeader + offset, sizeof(ulong));
+                            }
+                        } while (offset + sizeof(ulong) <= length);
+                    }
+                    if (offset + sizeof(uint) <= length)
                     {
-                        ThrowInvalidHeaderCharacter(ch);
+                        var stringUint = (uint*)(pHeader + offset);
+                        offset += sizeof(uint);
+                        if (((*stringUint | (*stringUint + 0x00010001u) | (*stringUint - 0x00200020u)) & 0xff80ff80u) != 0)
```

Comment:
Don't need to `*stringUint | ` as `(*stringUint - 0x00200020u)` should pickup the high sets